File: clojure/rest-euler/src/rest_euler/uceuler2.clj
(ns rest-euler.uceuler2
  "Use Case for problem id:2"
  (:gen-class))

(def fibonacci 
  (map first (iterate (fn [[a b]] [b (+ a b)]) [0 1])))

(defn even_fibonacci [] 
  (reduce + 
    (for [n fibonacci 
          :let [m2 (mod n 2)]
          :when (= m2 0)
          :while (< n 4000000)] n)))

(defn solve [id]
  (str "Solved UC " id ": Result is: " 
    (even_fibonacci)))

File: haskell/rest-euler/src/UCeuler2.hs
module UCeuler2
( solve
) where

-- With an anonymus function that takes a pair (a, b) in list and 
-- returns a new pair with (b (+ a b)).
-- We iterate over that function, i.e the output is the new input.
-- The first position in each pair is the Fibonacci sequence
-- that is map(ped) to a new lazy sequence
-- The $ operator is for avoiding parenthesis. Anything appearing 
-- after it will take precedence over anything that comes before.  
fibonacci :: [Integer]
fibonacci = map fst $ iterate (\(a,b) -> (b,a+b)) (0,1)

-- create new lazy list of even fibonacci numbers
even_fibonacci :: [Integer]
even_fibonacci = [ x | x <- fibonacci, (x `mod` 2==0) ]

solve :: Int -> [Char]
solve ucid = "Solved UC "++show(ucid)++": Result is: "++show(sum (takeWhile (<4000000) even_fibonacci))